gt/main.c

```diff
@@ -945,6 +946,9 @@ prep_notify_pkt(struct ggu_notify_pkt *ggu_pkt, struct gatekeeper_if *iface)
 		notify_ipv4->total_length = rte_cpu_to_be_16(
 			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
 
+		if (unlikely(!iface->ipv4_hw_cksum))
+			notify_ipv4->hdr_checksum = rte_ipv4_cksum(notify_ipv4);
+
```

Comment:
We can move lines `notify_ipv4->hdr_checksum = 0;`, `notify_pkt->ol_flags |= PKT_TX_IPV4;`, `if (likely(iface->ipv4_hw_cksum))`, and `notify_pkt->ol_flags |= PKT_TX_IP_CKSUM;` from `fill_notify_pkt_hdr()` to here to use `set_ipv4_checksum()`.

Reply:
I ended up moving the UDP checksum setup to `prep_notify_pkt()` to keep all of the checksumming code setup and offload flags together. I also moved the analogous IPv6 fields to `prep_notify_pkt()` for the same reason. Otherwise the code would have been split in a confusing manner between `fill_notify_pkt_hdr()` and `prep_notify_pkt()`.

Reply:
So I think I am actually going to not do what I mentioned above. if you move *all* of this code to `prep_notify_pkt()` like I did, it creates a gap between `fill_notify_pkt_hdr()` and `prep_notify_pkt()` where the code was expecting some values to have been set. For example, the UDP datagram length field was expecting `l3_len` to be set. There might be other assumptions in the code, or there may be more added in the future, so I would prefer not to move this code around.

Instead, I decided to use your suggestion, but also leave `notify_pkt->ol_flags |= PKT_TX_IPV4;` in `fill_notify_pkt_hdr()`. Even though this is redundant since it is also set again later in `set_ipv4_checksum()`, I left it here to keep the IPv4 and IPv6 cases parallel in `fill_notify_pkt_hdr()`.

Reply:
It reads like a reasonable plan.